Commit with new futures and fixes (fixes #7)

mkgerasimenko opened the pull request:
- Created custom assertions for Client and Panel entities;
 - Created an interface Reader for reading data from a custom data source;
 - Created concrete implementations for reading data from json and xml format;
 - Created a generic utility class for data retrieval;
 - Created several tests, which read data from different external data sources;
 - Added Service Provider Interface;

sskorol commented on build.gradle:
```diff
@@ -8,16 +12,40 @@ repositories {
+            "org.projectlombok:lombok:${lombokVersion}",
+            "org.assertj:assertj-core:${assertjVersion}",
+            "com.google.code.gson:gson:${gsonVersion}",
+            'com.fasterxml.jackson.dataformat:jackson-dataformat-xml:2.9.2',
```
do you want to move other versions to `ext` section as well?

sskorol commented on build.gradle:
```diff
@@ -8,16 +12,40 @@ repositories {
     jcenter()
 }
 
+ext{
```
seems like a missing whitespace... do you want to reformat code?

sskorol commented on build.gradle:
```diff
@@ -8,16 +12,40 @@ repositories {
 
 test {
+    testLogging {
+        showStandardStreams = false
```
that's redundant if you're not using it

sskorol commented on src/main/java/com/waverleysoftware/assertions/ClientAssert.java:
```diff
@@ -0,0 +1,34 @@
+
+    public ClientAssert hasClientName(final String clientName) {
+        isNotNull();
+        final String name = actual.getFirstName();
```
split methods calling with variables assignment to increase code readability...

sskorol commented on src/main/java/com/waverleysoftware/assertions/ClientAssert.java:
```diff
@@ -0,0 +1,34 @@
+        return this;
+    }
+
+    public ClientAssert hasMasterCode(final int conditionForAge) {
```
how `masterCode` is related to `age`?

sskorol commented on src/main/java/com/waverleysoftware/assertions/CustomAssertions.java:
```diff
@@ -0,0 +1,21 @@
+/**
+ * A simple base class for Assertions.
+ */
+@SuppressWarnings("Warning")
```
which warning are you suppressing?

sskorol commented on src/main/java/com/waverleysoftware/listemers/SecondTestListener.java:
```diff
@@ -0,0 +1,37 @@
+        log.info("Finish!!!!");
+    }
+
+    public static Reader getCurrentImpl(final String dataSource) {
```
give it more obvious name... what does "current impl" mean in your app's context?

sskorol commented on src/main/java/com/waverleysoftware/listemers/SecondTestListener.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    public static Reader getCurrentImpl(final String dataSource) {
+        final String extension = capitalize(dataSource.substring(dataSource.lastIndexOf('.')).replace(".", ""));
```
`commons-lang3` library which you are using, contains powerful string utils class, where you can get a substring after last character occurrence

sskorol commented on src/main/java/com/waverleysoftware/utils/ImplLoader.java:
```diff
@@ -0,0 +1,25 @@
+/**
+ * A simple service-provider loading facility.
+ */
+public final class ImplLoader {
```
make it more generic... it should be a service loader for any interface you provide

sskorol commented on src/main/java/com/waverleysoftware/listemers/SecondTestListener.java:
```diff
@@ -0,0 +1,37 @@
+    public static Reader getCurrentImpl(final String dataSource) {
+        final String extension = capitalize(dataSource.substring(dataSource.lastIndexOf('.')).replace(".", ""));
+        final Optional<Reader> readerOptional = StreamEx.of(projectImplementations)
+                .filter(impl -> impl.toString().contains(extension))
```
there's a method called `endsWith`... you may want to use it instead...

sskorol commented on src/main/java/com/waverleysoftware/listemers/SecondTestListener.java:
```diff
@@ -0,0 +1,37 @@
+        final String extension = capitalize(dataSource.substring(dataSource.lastIndexOf('.')).replace(".", ""));
+        final Optional<Reader> readerOptional = StreamEx.of(projectImplementations)
+                .filter(impl -> impl.toString().contains(extension))
+                .findFirst();
```
you can replace previous filter with conditional `findFirst`

sskorol commented on src/main/java/com/waverleysoftware/listemers/SecondTestListener.java:
```diff
@@ -0,0 +1,37 @@
+        final Optional<Reader> readerOptional = StreamEx.of(projectImplementations)
+                .filter(impl -> impl.toString().contains(extension))
+                .findFirst();
+        return readerOptional.orElseGet(readerOptional::get);
```
no need to save result into variable if you return it on next line...

sskorol commented on src/main/java/com/waverleysoftware/model/Client.java:
```diff
@@ -0,0 +1,21 @@
+import javax.xml.bind.annotation.XmlRootElement;
+
+/**
+ * A simple model class for panel processing.
```
it doesn't seem like a panel

sskorol commented on src/main/java/com/waverleysoftware/utils/ImplLoader.java:
```diff
@@ -0,0 +1,25 @@
+    public static List<Reader> getAllProjectImplementation() {
+        final List<Reader> implList = new ArrayList<>();
+        final ServiceLoader<Reader> serviceLoader = ServiceLoader.load(Reader.class);
+        for (Reader reader : serviceLoader) {
```
use streams instead

sskorol commented on src/main/java/com/waverleysoftware/utils/api/Reader.java:
```diff
@@ -0,0 +1,8 @@
+/**
+ * A simple interface for read of file processing.
+ */
+public interface Reader {
```
give it more obvious name

sskorol commented on src/test/java/com/waverleysoftware/AccessCodeTests.java:
```diff
@@ -1,32 +1,36 @@
-    public void accessCodeShouldBeCreated() {
-        System.out.println("Access code created");
+/**
+ * A simple model class for access code testing.
```
it doesn't seem like a model class

sskorol commented on src/test/java/com/waverleysoftware/AccessCodeTests.java:
```diff
@@ -1,32 +1,36 @@
+public class AccessCodeTests extends BaseTest {
+    @Test(dataProvider = "getPanelData", dataProviderClass = DataSuppliers.class)
+    public void accessCodeShouldBeAdded(final Panel... panels) {
+        final String conditionForType = "home";
```
it makes no sense to create a separate local variable, if you then use it on next like while assertion

sskorol commented on src/test/java/com/waverleysoftware/AccessCodeTests.java:
```diff
@@ -1,32 +1,36 @@
-        System.out.println("Access code deleted");
+    @Test(dataProvider = "getPanelData", dataProviderClass = DataSuppliers.class)
+    public void accessCodeShouldBeCreatedWithSomeDates(final Panel... panels) {
+        SOFTASSERT.assertThat(asList(panels).get(0).getName()).as("panel name").isEqualTo("lyn");
```
if you're always using first element, why don't you return just a single object from data supplier?

sskorol commented on src/test/java/com/waverleysoftware/BaseTest.java:
```diff
@@ -1,16 +1,42 @@
+    @BeforeMethod
+    public void setUp(final ITestContext context, final Method method) {
+        StreamEx.of(method.getDeclaredAnnotations())
+                .forEach(System.out::println);
```
it's not recommended to use `sout` in production code... replace it with logger

sskorol commented on src/test/java/com/waverleysoftware/DataFromFileTest.java:
```diff
@@ -0,0 +1,27 @@
+    @Data(source = "panel.json")
+    @Test(dataProvider = "getPanelDataFromFile", dataProviderClass = DataSuppliers.class)
+    public void readDataFromFileJson(final Panel panel) {
+        CustomAssertions.assertThat(panel).hasPanelName("adc");
```
use static import to avoid class specification before method call

sskorol reviewed (changes requested):
@mkgerasimenko check comments please

mkgerasimenko committed:
Added and changed the following:
 - refactor entities with new annotations;
 - added softAssertions;
 - created a custom annotation;
 - replaced common java logger with @Slf4j;
 - created custom assertions for Client and Panel entities;
 - created an interface Reader for reading data from a custom data source;
 - created concrete implementations for reading data from json and xml format;
 - created a generic utility class for data retrieval;
 - created several tests, which read data from different external data sources;
 - added Service Provider Interface;
 - moved of all dependencies versions to ext section;
 - make some methods more generic;
 - change loop to Stream;
 - added new packages;
 - added jOOR;
 - make some methods more generic;
 - added new parameter to custom annotation;
 - played with reflection;
 - common refactoring.

mkgerasimenko commented:
@sskorol, please look at the latest changes for this branch.

sskorol commented on src/test/java/com/waverleysoftware/data/DataSuppliers.java:
```diff
@@ -0,0 +1,50 @@
+
+    @SuppressWarnings("unchecked")
+    private <T> T getTypeByProvidedInfo(final Method method) {
+        final Data declaredAnnotation = method.getDeclaredAnnotation(Data.class);
```
you can put it into `ofNullable`

sskorol commented on src/main/java/com/waverleysoftware/assertions/ClientAssert.java:
```diff
@@ -0,0 +1,49 @@
+        return this;
+    }
+
+    public ClientAssert hasNextAge(final int clientAge) {
```
so what does `next` mean in this context?

sskorol reviewed (changes requested):
@mkgerasimenko check comments please

mkgerasimenko committed:
Changed the following:
 - put "declaredAnnotation" variable into ofNullable;
 - common refactoring.

sskorol reviewed (approved):
LGTM
